### PR TITLE
Make oneof fields to be the value directly

### PR DESCRIPTION
**Pull request by balusu-murali**
- oneof fields now must directly be the value.
- In addition to that, the variable type must match the record type.
- We extract variable name from the record name and then search for it.
- This will now always enforce the restriction that the variable name must match the message name in the protobuf definitions.

### CODE REVIEW

**Review by NikolaBorisov — APPROVED** [review]
Looks good.

ideally we would add an option to the one of definition where we can say that we don't want the tuples:
https://github.com/protocolbuffers/protobuf/blob/master/src/google/protobuf/descriptor.proto#L626

Also it would be good to check that there is no duplicate types and that they are all records. Both of those we can move to another change.

**Comment by balusu-murali** [comment]
> Looks good.
> 
> ideally we would add an option to the one of definition where we can say that we don't want the tuples:
> https://github.com/protocolbuffers/protobuf/blob/master/src/google/protobuf/descriptor.proto#L626
> 
> Also it would be good to check that there is no duplicate types and that they are all records. Both of those we can move to another change.

I checked that.. currently gpb ignores those options!
I will be adding an option to enif_protobuf.

**Review by mdonohue — APPROVED** [review]

**Comment by balusu-murali** [comment]
made some more changes. please take another look.

File: compile_commands.json [hunk 1]
```diff
@@ -0,0 +1,22 @@
+[
```
Comment (NikolaBorisov) [inline compile_commands.json]: what is this?

Comment (balusu-murali) [reply compile_commands.json]: removed it. some autogenerated file on compilation.

**Review by mdonohue — APPROVED** [review]

**Comment by NikolaBorisov** [comment]
LGTM

File: c_src/enif_protobuf.c [hunk 2]
```diff
@@ -20,6 +20,18 @@ make_atom(ErlNifEnv *env, const char *name)
     return enif_make_atom(env, name);
 }
 
+
+char*
+get_atom(ErlNifEnv *env, ERL_NIF_TERM term, char *buf, unsigned size)
+{
+    if (enif_get_atom(env, term, buf, size, ERL_NIF_LATIN1)) {
```
Comment (vipin211) [inline c_src/enif_protobuf.c:27]: question: why cann't the caller call enif_get_atom(,,,, ERL_NIF_LATIN1)?

File: c_src/ep_encoder.c [hunk 3]
```diff
@@ -617,13 +617,26 @@ get_oneof_field(ErlNifEnv *env, ERL_NIF_TERM term, ep_node_t *node, ERL_NIF_TERM
...
 
-    *out = array[1];
+    unsigned size = 100;
+    char *buf = _alloc(size);
+    char* output = get_atom(env, array[0], buf, size);
+
+    if (output == NULL) {
+        _free(buf);
+        return NULL;
+    } else {
```
Comment (vipin211) [inline c_src/ep_encoder.c]: you don't need the else

**Comment by balusu-murali** [comment]
@vipin211 can you please review this diff.

File: c_src/ep_encoder.c (hunk 3 above)
Comment (balusu-murali) [reply c_src/ep_encoder.c]: Ah.. yes, let me update it.

File: c_src/enif_protobuf.c (hunk 2 above)
Comment (balusu-murali) [reply c_src/enif_protobuf.c:27]: no reason actually..
caller also can call.. i felt it is in-line with the style of the repo..
would you like the other way better?

Comment (vipin211) [reply c_src/enif_protobuf.c:27]: what you have is fine.

**Review by vipin211 — APPROVED** [review]

**Pull request merged**